File: zenoh-grpc-client-sdk/zenoh-grpc-python/examples/sub.py
import zenoh_grpc
import time

with zenoh_grpc.Session.connect() as session:
    with session.declare_subscriber("demo/example/**") as sub:
        for i in range(1000):
            sample = sub.recv() 
            # blocking until a sample is received.
            # not recommended to use recv() in a loop, as it is blocking and will block the main thread, and due to the inner rust implementation, it can not be killed by Ctrl+C. It is better to use a callback to receive samples, as shown in sub_callback.py. But here we just want to show how to use recv() in a loop, so we use it here for simplicity.

            print(sample.key_expr, sample.payload, sample.encoding)
            # print("dropped count:", sub.dropped_count())
            # time.sleep(1) # do not sleep here, just for testing the dropped count (The low consumption speed leads to the inner subscription queue dropping data.)

File: zenoh-grpc-client-sdk/zenoh-grpc-cpp/examples/sub.cpp
#include <iostream>

#include "common.hpp"

int main(int argc, char** argv) {
    auto session = zenoh_grpc::Session::connect(example_endpoint(argc, argv));

    zenoh_grpc::Session::SubscriberOptions options;
    options.allowed_origin = zenoh_grpc::Locality::Any;
    auto subscriber = session.declare_subscriber("demo/example/**", options);

    std::cout << "subscriber ready on " << example_endpoint(argc, argv) << std::endl;
    for (int i = 0; i < 500; ++i) {
        auto sample = subscriber.recv();
        print_sample(sample);
        std::cout << "dropped=" << subscriber.dropped_count() << std::endl;
    }

    return 0;
}
